Software/Development/SemesterProject/App/referenceHandler.c

```diff
@@ -0,0 +1,317 @@
+    {
+        torqueReference = 0;
+        speedReference = 0;
+        if ((readSystemState() == RUNNING)) referenceSourceChanged = 1; //If the referenceSOurce is changed while in running, qualify the change so the system know it must go back to STANDBY
+        return 1;
+    }
+    else    return 0;
+}
+
+int referenceTypeIsChanged(void)
+{
+    return (GUISignals.ReferenceTypePushbutton || referenceTypeHasBeenPressed());
+}
+
+/*
+ * This FSM analyses what is the reference source selected and then acquires
+ * the torque reference from Analog read or graphical user interface.
+ */
+float getReferenceValue(void)
+{
+    float analogValueTorque = 0, torqueReferenceBeforeLimit = 0;
+    switch (referenceSource)
+    {
+    case interfacePCB:
+    {
+        analogValueTorque = getTorqueReferenceSliderMeasurement();
+        torqueReferenceBeforeLimit = CalculateTorqueRefFromAnalog(analogValueTorque);
+    }break;
+    case goKart:
+    {
+        analogValueTorque = getTorqueReferencePedalMeasurement();
+        torqueReferenceBeforeLimit = CalculateTorqueRefFromAnalog(analogValueTorque);
+    }break;
+    case GUI:
+    {
+        torqueReferenceBeforeLimit = GUISignals.TorqueReference;
+    }break;
+    }
+    return torqueReferenceBeforeLimit;
+}
+
+float CalculateTorqueRefFromAnalog(float analogRead)
+{
+    return analogRead/MAX_VALUE_ADC*MAX_TORQUE_REF_NM;
+}
+
+float calculateLimitedTorqueReference(float torqueReferenceBeforeLimit)
+{
+    if (referenceIncreased(torqueReferenceBeforeLimit)) return torqueReference + deltaTorque;
+    if (referenceDecreased(torqueReferenceBeforeLimit)) return torqueReference - deltaTorque;
+    else    return torqueReference;
+}
+
+int referenceIncreased(float torqueReferenceBeforeLimit)
+{
+    return torqueReferenceBeforeLimit > torqueReference + deltaTorque;
+}
+
+int referenceDecreased(float torqueReferenceBeforeLimit)
+{
+    return torqueReferenceBeforeLimit < torqueReference - deltaTorque;
+}
+
+float calculateSaturatedTorque(float torqueBeforeSaturation)
+{
+    if (torqueBeforeSaturation > MAX_TORQUE_REF_NM) return MAX_TORQUE_REF_NM;
+    else return torqueBeforeSaturation;
+}
+
+void calculateSpeedReference(void)
+{
+    int speedReferenceBeforeLimit = 0, speedReferenceBeforeSaturation = 0;
+    speedReferenceBeforeLimit = getSpeedReferenceValue();
+    speedReferenceBeforeSaturation = calculateLimitedSpeedReference(speedReferenceBeforeLimit);
+    speedReference = calculateSaturatedSpeed(speedReferenceBeforeSaturation);
+}
+int getSpeedReferenceValue(void)
+{
+    if (speedMustBeIncreased())     undampedSpeedReference += SPEED_STEP_FROM_BUTTON_RPM;
+    if (speedMustBeDecreased())     undampedSpeedReference -= SPEED_STEP_FROM_BUTTON_RPM;
+
+    return undampedSpeedReference;
+}
+
+int calculateLimitedSpeedReference(int speedReferenceBeforeLimit)
+{
+    if (speedRefIncreased(speedReferenceBeforeLimit)) return speedReferenceBeforeLimit + deltaSpeed;
```

Comment:
Having the argument in the function makes it less understandable for me, i think it would be better to have if (speedRefIncreased()) return...
in any case the variable speedRefBeforeLimit is the same used in the function anyway.

Reply:
The thing is that this variable is not a global, then that can't be done. Added function header for explaining.

Reply:
Ok, makes sense.